### PR TITLE
Elevator

### PR DESCRIPTION
**Pull request by mahirsowad3**
Made constraints that determines when the elevator sensor is at predefined heights and then prints rumbling to logger.  Later, planning on using for making controller rumble when passing these reference heights.

### CODE REVIEW

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	private static ElevatorSensor instance;
+	private static final Logger LOGGER = Logger.getLogger(ElevatorSensor.class);
+
+	private static final double TICKS_PER_TURN = 253.0;
+	private static final double GEAR_CIRCUMFERENCE_IN_INCHES = 10;	
+	public static final double ELEVATOR_MAX_HEIGHT_IN_FEET = 10;
+	public static final double ELEVATOR_MIN_HEIGHT_IN_FEET = 0;
+	public static final double ELEVATOR_INITIAL_TICKS = 196;
+	public static final int ELEVATOR_HEIGHT_SENSOR_ID = 0;
+	public static final int TALON_HEIGHT_CONTROLLER_ID = 1;
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: Put these values in the RobotMap.

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		
+		Stops(double height) {
+			this.height = height;
+		}
+	}
+	
+	private AnalogInput heightSensor;
+	private WPI_TalonSRX heightController;
+	private double feetPerTick;
+	private double previousHeight;
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: Put the enum after all the variables.

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		if (instance == null) {
+			instance = new ElevatorSensor();
+		}
+		return instance;
+	}
+
+	/**
+	 * Moves based on the Xbox controller analog input
+	 * 
+	 * @param speed the speed and direction. Shall be a value between -1 and 1.
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: The speed AND direction? Where's the direction?

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	 */
+	public void manualMove(double speed) {
+		if (isOutOfRange()) {
+			speed = 0;
+			// TODO: Rumble here
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: If you do this in the main Robot class, no need to have it here. And vice versa.

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		}
+		double currentHeight = getHeight();
+		for (Stops stop : Stops.values()) {
+			if ((previousHeight < stop.height && currentHeight >= stop.height)
+				|| (previousHeight > stop.height && currentHeight <= stop.height)) {
+				
+				//multiple prints are easier to spot on logger
+				LOGGER.debug("----- Rumbling -----");
+			}
+
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: Delete excess whitespace.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 6]
```diff
@@ -10,9 +10,17 @@
 //import edu.wpi.first.wpilibj.networktables.NetworkTable;
 import edu.wpi.cscore.UsbCamera;
 import edu.wpi.first.wpilibj.CameraServer;
-import edu.wpi.first.wpilibj.TimedRobot;
+
+import edu.wpi.first.wpilibj.IterativeRobot;
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: Remove this import if you don't use it.

**Review by npetrangelo — CHANGES_REQUESTED** [review]

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java (hunk 2 above)
Comment (AndrewItIs) [inline src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: Try to place as many constants up into the RobotMap class

File: src/org/usfirst/frc/team467/robot/ElevatorSensor.java (hunk 3 above)
Comment (npetrangelo) [reply src/org/usfirst/frc/team467/robot/ElevatorSensor.java]: Oh, never mind, it's 1-dimensional movement. But you should specify which sign means up and which sign means down.

File: src/org/usfirst/frc/team467/robot/Drive.java [hunk 7]
```diff
@@ -61,6 +61,7 @@ private Drive(WPI_TalonSRX leftLead,  WPI_TalonSRX leftFollower1,  WPI_TalonSRX
 		
 		this.rightLead = rightLead;
 		initMotor(this.rightLead);
+
```
Comment (AndrewItIs) [inline src/org/usfirst/frc/team467/robot/Drive.java]: Try to keep the rightLead stuff with the right lead stuff and delete anything unused.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 8]
```diff
@@ -10,9 +10,17 @@
...
+import edu.wpi.first.wpilibj.networktables.NetworkTable;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 
+import org.usfirst.frc.team467.robot.Autonomous.ActionGroup;
+// import org.usfirst.frc.team467.robot.Autonomous.Actions;
```
Comment (AndrewItIs) [inline src/org/usfirst/frc/team467/robot/Robot.java]: delete this commented Autonomous actions import if you import it the very next line.

**Review by npetrangelo — COMMENTED** [review]
Looks good, but I think we should wait to merge until we have an elevator to test it on.

**Comment by AndrewItIs** [comment]
like Nate said, looks fine, but we have to test it first

File: src/org/usfirst/frc/team467/robot/Drive.java (hunk 7 above)
Comment (mahirsowad3) [reply src/org/usfirst/frc/team467/robot/Drive.java]: What do you mean by this?

Comment (AndrewItIs) [reply src/org/usfirst/frc/team467/robot/Drive.java]: the set inverted and put the rightlead init and theSetSensorPhase together

File: .classpath [hunk 9]
```diff
@@ -2,15 +2,16 @@
 <classpath>
 	<classpathentry kind="src" path="src"/>
 	<classpathentry kind="var" path="wpilib" sourcepath="wpilib.sources"/>
+	<classpathentry kind="var" path="networktables" sourcepath="networktables.sources"/>
 	<classpathentry kind="var" path="opencv" sourcepath="opencv.sources"/>
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/log4j-1.2.17.jar"/>
-	<classpathentry kind="var" path="wpiutil" sourcepath="wpiutil.sources"/>
```
Comment (npetrangelo) [inline .classpath:9]: Add this one back

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	private double feetPerTick;
+	private int maxTicksPerIteration;
+	private double previousHeight;
+	private MotorSafetyHelper m_safetyHelper = new MotorSafetyHelper(heightController);
+	
+	private Stops targetHeight;
+	
+	public enum Stops {
+		noStop(RobotMap.ELEVATOR_MIN_HEIGHT_IN_FEET),
+		min(RobotMap.ELEVATOR_MIN_HEIGHT_IN_FEET),
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: No need for both of these

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	 * 1 rotation=253 ticks
+	 * 
+	 * @return a single instance of the Elevator object.
+	 */
+	public static Elevator getInstance() {
+		if (instance == null) {
+			instance = new Elevator();
+		}
+		return instance;
+	}
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Put getInstance() before constructor

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+		
+		double currentHeight = getHeight();
+		for (Stops stop : Stops.values()) {
+			if ((previousHeight < stop.height && currentHeight >= stop.height)
+				|| (previousHeight > stop.height && currentHeight <= stop.height)) {
+				
+				//DriverStation.getInstance().setDriverRumble(0.5);
+			} else {
+				//DriverStation.getInstance().setDriverRumble(0.0);
+			}
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Fix the spacing

File: .classpath [hunk 13]
```diff
@@ -2,15 +2,16 @@
...
 	<classpathentry kind="var" path="USERLIBS_DIR/CTRE_Phoenix-javadoc.jar"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/CTRE_Phoenix-sources.jar"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/CTRE_Phoenix.jar"/>
-	<classpathentry kind="var" path="networktables" sourcepath="networktables.sources"/>
+	<classpathentry kind="var" path="USERLIBS_DIR/ntcore-java-4.0.0.jar"/>
+	<classpathentry kind="var" path="USERLIBS_DIR/ntcore-jni-4.0.0-all.jar"/>
```
Comment (john-petrangelo) [inline .classpath]: I think these are duplicate lines, remove one copy.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 14]
```diff
@@ -36,6 +43,10 @@
 	private ActionGroup autonomous;
 	
 	private Gyrometer gyro;
+
+
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: Too many blank lines, please remove one.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+		targetHeight = Stops.noStop;
+		if (m_safetyHelper != null) {
+			m_safetyHelper.feed();
+		}
+		if (isOutOfRange()) {
+			heightController.set(0);
+			//DriverStation.getInstance().setDriverRumble(0.5);
+			return; // Don't bother with any more logic here.
+		}
+
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Set previous height to current height before setting current height

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 16]
```diff
@@ -100,7 +117,11 @@ public void disabledPeriodic() {
 //	
 	public void autonomousInit() {
 		final String autoMode = SmartDashboard.getString("Auto Selector", "none");
+
+		//drive.initMotionMagicMode();
 		
+//	drive.initPositionMode(); 
+
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: Why did you add this commented out code? Probably should be removed.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 17]
```diff
@@ -138,7 +161,6 @@ public void testPeriodic() {
 			LOGGER.info("You pressed b");
 		}
 
-		driverstation.periodic();
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java:148]: Don't remove this.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+		LOGGER.debug("Current Height: " + currentHeight);
+		heightController.set(speed);
+		previousHeight = currentHeight;
+		LOGGER.debug("Previous Height: " + previousHeight);
+	}
+	
+	public void initMotionMagicMode() {
+//		heightController.setSelectedSensorPosition(0, 0, RobotMap.TALON_TIMEOUT);
+
+
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Remove excess whitespace

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+			//DriverStation.getInstance().setDriverRumble(0.5);
+			return; // Don't bother with any more logic here.
+		}
+		
+		double currentHeight = getHeight();
+		for (Stops stop : Stops.values()) {
+			if ((previousHeight < stop.height && currentHeight >= stop.height)
+				|| (previousHeight > stop.height && currentHeight <= stop.height)) {
+				
+				//DriverStation.getInstance().setDriverRumble(0.5);
```
Comment (npetrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Mark these with TODOs mentioning the Rumblers

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 20]
```diff
@@ -168,6 +190,10 @@ public void teleopPeriodic() {
 	 		right = 0.0;
 	 	}
 	 	
+		double ElevatorScale = driverstation.getNavJoystick().getLeftStickY();
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: I think a better name instead of elevatorScale would be elevatorSpeed.

Also, variables should always start with lowercase letters.

File: src/org/usfirst/frc/team467/robot/RobotMap.java [hunk 21]
```diff
@@ -12,6 +18,18 @@
 	public static final int BACK_LEFT = 2;
 	public static final int BACK_RIGHT = 3;
 	
+	public static final int ELEVATOR_TICKS_PER_TURN = 253;
+	public static final double GEAR_CIRCUMFERENCE_IN_INCHES = 10;
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/RobotMap.java]: Which gear? If this is an elevator gear then prefix it with ELEVATOR_.

File: src/org/usfirst/frc/team467/robot/RobotMap.java [hunk 22]
```diff
@@ -12,6 +18,18 @@
...
+	public static final double ELEVATOR_MAX_HEIGHT_IN_FEET = 10;
+	public static final double ELEVATOR_MIN_HEIGHT_IN_FEET = 0;
+	public static final int ELEVATOR_INITIAL_TICKS = 196;
+	public static final int ELEVATOR_HEIGHT_SENSOR_ID = 0;
+	public static final int ELEVATOR_MOTOR_CHANNEL = 1;
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/RobotMap.java]: I think you should not set a motor channel here, it varies with every robot.

**Review by npetrangelo — CHANGES_REQUESTED** [review]
Minor changes need, but overall this looks pretty good.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+			this.height = height;
+		}
+	}
+
+	private Elevator() {
+		heightSensor = new AnalogInput(RobotMap.ELEVATOR_HEIGHT_SENSOR_ID);
+		heightController = new WPI_TalonSRX(RobotMap.ELEVATOR_MOTOR_CHANNEL);
+		targetHeight = Stops.noStop;
+		feetPerTick = (RobotMap.GEAR_CIRCUMFERENCE_IN_INCHES / 12) / RobotMap.ELEVATOR_TICKS_PER_TURN;
+		maxTicksPerIteration = RobotMap.ELEVATOR_TICKS_PER_TURN * RobotMap.MAX_ELEVATOR_RPM / 60 / 100; // 10 ms per iteration
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: I don't think maxTicksPerIteration should be an instance variable, just calculate it inside the initMotionMagicMode() method.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+		previousHeight = getHeight();
+	}
+	
+	/**
+	 * The lowest value is 196.0, the maximum value is 3741.0. The middle is 1968.5
+	 * New max: 2980, new min:956.5
+	 * 16.9 ticks = 1 inch
+	 * 1 rotation=253 ticks
+	 *
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: These number comments don't really belong here, they are not used in the nearby code. Have they been moved to RobotMap now?

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+				
+				//DriverStation.getInstance().setDriverRumble(0.5);
+			} else {
+				//DriverStation.getInstance().setDriverRumble(0.0);
+			}
+		}
+		if(currentHeight >= 2 && currentHeight <= 2.1 || currentHeight >= 4 && currentHeight <=4.1 || currentHeight >= 6 && currentHeight <= 6.1)
+			LOGGER.debug("Rumbling right now");
+			else {		
+			}
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Is this "if" block of code obsolete now, because it does almost the same thing as the for loop with stops above?

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+		LOGGER.debug("Current Height: " + currentHeight);
+		heightController.set(speed);
+		previousHeight = currentHeight;
+		LOGGER.debug("Previous Height: " + previousHeight);
+	}
+	
+	public void initMotionMagicMode() {
+//		heightController.setSelectedSensorPosition(0, 0, RobotMap.TALON_TIMEOUT);
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Don't keep commented out code like this, if you don't need it then remove it.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+
+		
+		double kPElevator = 1.4; // Double.parseDouble(SmartDashboard.getString("DB/String 7", "1.4"));
+		double kIElevator = 0.0; // Double.parseDouble(SmartDashboard.getString("DB/String 8", "0.0"));
+		double kDElevator = 165; // Double.parseDouble(SmartDashboard.getString("DB/String 9", "165"));
+		double kFElevator = 0.5; //  Double.parseDouble(SmartDashboard.getString("DB/String 6", "0.5"));
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: Should should be in RobotMap.

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+		
+		heightController.config_kP(0, kPElevator, RobotMap.TALON_TIMEOUT);
+		heightController.config_kI(0, kIElevator, RobotMap.TALON_TIMEOUT);
+		heightController.config_kD(0, kDElevator, RobotMap.TALON_TIMEOUT);
+		heightController.config_kF(0, kFElevator, RobotMap.TALON_TIMEOUT);
+
+		//		This is commented out because we will need the SmartDashboard to tune other things later.
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: What is commented out?

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+		
+		heightController.configMotionCruiseVelocity(maxTicksPerIteration / 2, RobotMap.TALON_TIMEOUT);
+		heightController.configMotionAcceleration(maxTicksPerIteration / 2, RobotMap.TALON_TIMEOUT);	
+	}
+
+	public boolean isOutOfRange() {
+		return (getHeight() > RobotMap.ELEVATOR_MAX_HEIGHT_IN_FEET || getHeight() < RobotMap.ELEVATOR_MIN_HEIGHT_IN_FEET);
+	}
+
+	public double getHeight() {
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: I'd like getHeight() to be renamed to getHeightFeet().

File: src/org/usfirst/frc/team467/robot/Elevator.java [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+		double height = (heightSensor.getValue() - RobotMap.ELEVATOR_INITIAL_TICKS) * feetPerTick;
+		LOGGER.debug("Height in feet: " + height);
+		return height;
+	}
+	
+	public void targetHeight(Stops target) {
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Elevator.java]: I think all of the "height" related methods should include the units in their name. Otherwise you could confuse feet with inches and crash your probe into the surface of the planet.

**Review by john-petrangelo — CHANGES_REQUESTED** [review]

File: .classpath [hunk 31]
```diff
@@ -7,12 +7,12 @@
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/log4j-1.2.17.jar"/>
+    <classpanthetry kind ="var" path ="wpiutil" sourcepath="wpiutil.sources"/>
```
Comment (john-petrangelo) [inline .classpath]: This is indented funny..

File: src/org/usfirst/frc/team467/robot/Elevator.java (hunk 19 above)
Comment (john-petrangelo) [reply src/org/usfirst/frc/team467/robot/Elevator.java]: Still missing the TODOs.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 32]
```diff
@@ -58,68 +57,65 @@
...
+
 		//made usb camera and captures video
-		//UsbCamera cam = CameraServer.getInstance().startAutomaticCapture();
+		UsbCamera cam = CameraServer.getInstance().startAutomaticCapture();
 		//set resolution and frames per second to match driverstation
-		//cam.setResolution(320, 240);
-		//cam.setFPS(15);
-		
+		cam.setResolution(320, 240);
+		cam.setFPS(15);
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: I'd rather leave the camera commented out for now. Otherwise we crash looking for cameras that don't exist.

File: src/org/usfirst/frc/team467/robot/Robot.java [hunk 33]
```diff
@@ -137,15 +133,15 @@ public void autonomousInit() {
...
-//		autonomous.terminate();
-//		autonomous = Actions.doNothing();
+		//		autonomous.terminate();
+		//		autonomous = Actions.doNothing();
+		driverstation.periodic();
 	}
 
 	public void testInit() {
-		elevator.targetHeight(Stops.fieldSwitch);
+		elevator.moveToHeight(Stops.fieldSwitch);
```
Comment (john-petrangelo) [inline src/org/usfirst/frc/team467/robot/Robot.java]: Names like this should include the units in the name.

**Review by npetrangelo — APPROVED** [review]

**Review by john-petrangelo — APPROVED** [review]

**Review by john-petrangelo — APPROVED** [review]
